File: Университет/cp_trainings/СНМ (DSU)/ITMO CF/step1/A/A.py
class DSU:
    def __init__(self, n: int) -> None:
        self.p = [i for i in range(n)]
        self.size = [1] * n
        return
    
    def get(self, a: int) -> int:
        if self.p[a] != a:
            self.p[a] = self.get(self.p[a])
        return self.p[a]
    
    def union(self, a: int, b: int) -> None:
        a = self.get(a)
        b = self.get(b)
        if a == b:
            return
        if self.size[a] < self.size[b]:
            a, b = b, a
        self.p[b] = a
        self.size[a] += self.size[b]
        return

def main() -> None:
    n, q = map(int, input().split())
    dsu = DSU(n)
    for _ in range(q):
        cmd, u, v = input().split()
        u, v = int(u) - 1, int(v) - 1
        if cmd == 'union':
            dsu.union(u, v)
        else: # cmd == 'get'
            print('YES' if dsu.get(u) == dsu.get(v) else 'NO')
    return

if __name__ == '__main__':
    main()

File: Университет/cp_trainings/СНМ (DSU)/ITMO CF/step1/A/A.cpp
// author: Danila "akshin_" Axyonov

#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
using namespace std;
typedef long long int lli;

#define plli                pair<lli, lli>
#define mp(_first, _second) make_pair(_first, _second)
#define pb(_elem)           push_back(_elem)




class DSU {
private:
    vector<lli> p,
                size;

public:
    DSU(lli n) {
        this->p.resize(n);
        generate(this->p.begin(), this->p.end(), []() { static lli i = 0LL; return i++; });
        this->size.assign(n, 1LL);
        return;
    }

    lli get(lli a) {
        if (this->p[a] != a)
            this->p[a] = this->get(this->p[a]);
        return this->p[a];
    }

    void union_(lli a, lli b) {
        a = this->get(a);
        b = this->get(b);
        if (a == b)
            return;
        if (this->size[a] < this->size[b])
            swap(a, b);
        this->p[b] = a;
        this->size[a] += this->size[b];
        return;
    }
};




void solve();

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    
    solve();
    
    return 0;
}

void solve() {
    lli n,
        q;
    cin >> n >> q;
    DSU dsu(n);
    while (q--) {
        string cmd;
        lli u, v;
        cin >> cmd >> u >> v;
        --u, --v;
        if (cmd == "union")
            dsu.union_(u, v);
        else // cmd == "get"
            cout << (dsu.get(u) == dsu.get(v) ? "YES" : "NO") << '\n';
    }
    return;
}
